### PR TITLE
Extend deserialiseFromRawBytesHex to produce error description

### PR DESCRIPTION
**Pull request by cblp**

### CODE REVIEW

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Right, let's reorganize these commits:
Commit 1: Change to `deserialiseFromRawBytesHex`
Commit 2: Propagating this change throughout the repo
Commit 3: Miscellaneous refactorings

**Comment by cblp** [comment]
> Right, let's reorganize these commits: Commit 1: Change to `deserialiseFromRawBytesHex` Commit 2: Propagating this change throughout the repo Commit 3: Miscellaneous refactorings

done

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Right direction, a few comments

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs [hunk 1]
```diff
@@ -32,8 +31,7 @@ serialiseToRawBytesHexText :: SerialiseAsRawBytes a => a -> Text
 serialiseToRawBytesHexText = Text.decodeUtf8 . serialiseToRawBytesHex
 
 deserialiseFromRawBytesHex :: SerialiseAsRawBytes a
-                           => AsType a -> ByteString -> Maybe a
-deserialiseFromRawBytesHex proxy hex =
-    case Base16.decode hex of
-      Right raw -> deserialiseFromRawBytes proxy raw
-      Left _msg -> Nothing
+                           => AsType a -> ByteString -> Either String a
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseRaw.hs:34]: I think it should be done in another patch. String is used already everywhere.

File: cardano-api/src/Cardano/Api/Utils.hs [hunk 2]
```diff
@@ -46,7 +45,8 @@ runParsecParser parser input =
     Right txin -> pure txin
     Left parseError -> fail $ formatParsecError parseError
 
-note :: MonadFail m => String -> Maybe a -> m a
-note msg = \case
-  Nothing -> fail msg
-  Just a -> pure a
+hoistEither :: MonadFail m => Either String a -> m a
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/Utils.hs:48]: yes, it's better!

File: cardano-cli/src/Cardano/CLI/Helpers.hs [hunk 3]
```diff
@@ -126,3 +127,6 @@ hushM :: forall e m a. Monad m => Either e a -> (e -> m ()) -> m (Maybe a)
 hushM r f = case r of
   Right a -> return (Just a)
   Left e -> f e >> return Nothing
+
+fmapL :: (a -> b) -> Either a c -> Either b c
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Helpers.hs]: oh, sure!

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/SerialiseRaw.hs:34]: > String is used already everywhere.

And it shouldn't be. Even more reason to change the error type.

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs [hunk 4]
```diff
@@ -31,9 +32,25 @@ serialiseToRawBytesHex = Base16.encode . serialiseToRawBytes
...
+  | RawBytesHexErrorRawBytesDecodeFail
+      ByteString -- ^ original input
+      -- TODO TypeRep -- ^ output type proxy
+  deriving (Show)
+
+instance Error RawBytesHexError where
+  displayError = \case
+    RawBytesHexErrorBase16DecodeFail input message ->
+      "Base16 cannot decode " ++ show input ++ ": " ++ message
+    RawBytesHexErrorRawBytesDecodeFail input ->
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/SerialiseRaw.hs:48]: `"Failed to deserialise: " ++ show input"`

In future (not this PR) we will want to also return a typed error for `SerialiseAsRawBytes`

File: bench/tx-generator/src/Cardano/Benchmarking/PlutusExample.hs [hunk 5]
```diff
@@ -62,8 +62,8 @@ readScript fp = do
 toScriptHash :: String -> Hash ScriptData
 toScriptHash str
   = case deserialiseFromRawBytesHex (AsHash AsScriptData) (BSC.pack str) of
-    Just x -> x
-    Nothing  -> error $ "Invalid datum hash: " ++ show str
+    Right x -> x
```
Comment (Jimbo4350) [inline bench/tx-generator/src/Cardano/Benchmarking/PlutusExample.hs:65]: Two spaced indent needed here

File: cardano-api/src/Cardano/Api/Script.hs [hunk 6]
```diff
@@ -97,47 +97,38 @@ module Cardano.Api.Script (
 
 import           Prelude
 
+import           Control.Applicative
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Script.hs:100]: :+1:

File: cardano-api/src/Cardano/Api/Script.hs [hunk 7]
```diff
@@ -1301,7 +1294,8 @@ parseScriptAfter lang =
         _       -> fail "\"after\" script value not found"
 
 parsePaymentKeyHash :: Text -> Aeson.Parser (Hash PaymentKey)
-parsePaymentKeyHash txt =
-    case deserialiseFromRawBytesHex (AsHash AsPaymentKey) (Text.encodeUtf8 txt) of
-      Just payKeyHash -> return payKeyHash
-      Nothing -> fail $ "Error deserialising payment key hash: " <> Text.unpack txt
+parsePaymentKeyHash =
+  failEitherWith
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Script.hs:1298]: :+1:

File: cardano-api/src/Cardano/Api/ValueParser.hs [hunk 8]
```diff
@@ -115,16 +116,17 @@ decimal = do
...
 
 -- | Policy ID parser.
 policyId :: Parser PolicyId
 policyId = do
   hexText <- many1 hexDigit
   case textToPolicyId hexText of
-    Just p -> pure p
-    Nothing ->
+    Right p -> pure p
+    Left _ ->
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/ValueParser.hs:129]: Why not use `displayError` here as well? Also are policy id's meant to only be 56 hex digits long?

File: cardano-cli/src/Cardano/CLI/Byron/Parsers.hs [hunk 9]
```diff
@@ -758,4 +758,3 @@ parseSigningKeyFile opt desc = SigningKeyFile <$> parseFilePath opt desc
 parseGenesisFile :: String -> Parser GenesisFile
 parseGenesisFile opt =
   GenesisFile <$> parseFilePath opt "Genesis JSON file."
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Byron/Parsers.hs:760]: Please undo the setting in your editor that automatically removes newlines from files. It's unnecessary diff.

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 10]
```diff
@@ -17,43 +17,42 @@ module Cardano.CLI.Shelley.Parsers
...
 import qualified Data.IP as IP
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Set as Set
 import qualified Data.Text as Text
 import qualified Data.Text.Encoding as Text
-
-import qualified Data.Aeson as Aeson
-import qualified Data.Aeson.Parser as Aeson.Parser
-import qualified Data.Attoparsec.ByteString.Char8 as Atto
+import           Data.Time.Clock (UTCTime)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:31]: Nice :+1:

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 11]
```diff
@@ -2264,24 +2266,24 @@ pStakePoolVerificationKeyFile =
...
+    pHexStakePoolId :: ReadM (Hash StakePoolKey)
     pHexStakePoolId =
-      deserialiseFromRawBytesHex (AsHash AsStakePoolKey) . BSC.pack
+      Opt.eitherReader $
+        first displayError
+          . deserialiseFromRawBytesHex (AsHash AsStakePoolKey)
+          . BSC.pack
 
-    pBech32StakePoolId :: String -> Maybe (Hash StakePoolKey)
+    pBech32StakePoolId :: ReadM (Hash StakePoolKey)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:2284]: :+1:

**Review by Jimbo4350 — COMMENTED** [review]
Looking good @cblp :smile: . A few minor fixes.

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs [hunk 12]
```diff
@@ -31,9 +32,27 @@ serialiseToRawBytesHex = Base16.encode . serialiseToRawBytes
...
+instance Error RawBytesHexError where
+  displayError = \case
+    RawBytesHexErrorBase16DecodeFail input message ->
+      "Expected Base16-encoded bytestring, but got " ++ show input ++ "; "
+      ++ message
+    RawBytesHexErrorRawBytesDecodeFail input ->
+      "Failed to deserialise " ++ show input
+      -- TODO(2022-01-26, cblp) show expected output type
+
+deserialiseFromRawBytesHex ::
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/SerialiseRaw.hs:53]: If you have a look in the code base you'll see we usually indent type signatures that break the 100 char limit as follows:
```
deserialiseFromRawBytesHex 
  :: SerialiseAsRawBytes a 
  => AsType a 
  -> ByteString 
  -> Either RawBytesHexError
```
Although in this case because we are under 100 chars, we don't need any indentation necessarily. Either or is fine.

**Review by Jimbo4350 — APPROVED** [review]

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs (hunk 12 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseRaw.hs:53]: I also found some files using Ormolu style, so I applied it here. Okay, looks like this file should be in classic style.

Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseRaw.hs]: The style guide says the (soft) limit is 80 chars.

**Comment by cblp** [comment]
bors merge

**Comment by cblp** [comment]
bors berge

**Comment by cblp** [comment]
bors merge

**Comment by cblp** [comment]
bors merge

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs [hunk 13]
```diff
@@ -31,9 +32,28 @@ serialiseToRawBytesHex = Base16.encode . serialiseToRawBytes
...
-deserialiseFromRawBytesHex :: SerialiseAsRawBytes a
-                           => AsType a -> ByteString -> Maybe a
-deserialiseFromRawBytesHex proxy hex =
-    case Base16.decode hex of
-      Right raw -> deserialiseFromRawBytes proxy raw
-      Left _msg -> Nothing
+data RawBytesHexError
+  = RawBytesHexErrorBase16DecodeFail
+      ByteString -- ^ original input
+      String -- ^ error message
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/SerialiseRaw.hs:38]: Can we not do better than a String here?

File: cardano-api/src/Cardano/Api/Script.hs [hunk 14]
```diff
@@ -104,60 +104,53 @@ module Cardano.Api.Script (
...
-import qualified Cardano.Ledger.Alonzo.Scripts as Alonzo
-
 import qualified Plutus.V1.Ledger.Examples as Plutus
 
 import           Cardano.Api.Eras
-import           Cardano.Api.Error
-import           Cardano.Api.HasTypeProxy
+import           Cardano.Api.Error (displayError)
 import           Cardano.Api.Hash
+import           Cardano.Api.HasTypeProxy
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Script.hs:145]: Do you really need to reorder all these imports? These are changes unrelated to the purpose of the PR.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 15]
```diff
@@ -428,10 +430,10 @@ instance FromJSONKey TxIn where
 
 parseTxId :: Parsec.Parser TxId
 parseTxId = do
-  str <- Parsec.many1 Parsec.hexDigit Parsec.<?> "transaction id (hexadecimal)"
-  case deserialiseFromRawBytesHex AsTxId (BSC.pack str) of
-    Just addr -> return addr
-    Nothing -> fail $ "Incorrect transaction id format:: " ++ show str
+  str <- some Parsec.hexDigit <?> "transaction id (hexadecimal)"
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/TxBody.hs:433]: Was there a reason to switch `many1` to `some`? The Parsec version has an implementation specialised to parsers.

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 16]
```diff
@@ -1974,7 +1975,7 @@ parseTxIn = TxIn <$> parseTxId <*> (Parsec.char '#' *> parseTxIx)
 
 parseTxId :: Parsec.Parser TxId
 parseTxId = do
-  str <- Parsec.many1 Parsec.hexDigit Parsec.<?> "transaction id (hexadecimal)"
+  str <- some Parsec.hexDigit <?> "transaction id (hexadecimal)"
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:1978]: Same question. Is this change needed? Why?

File: cardano-api/src/Cardano/Api/SerialiseRaw.hs (hunk 13 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseRaw.hs:38]: Sorry? Too much indirection for my English.

Do you suggest having something more structured than a String in the second field of RawBytesHexErrorBase16DecodeFail or not?

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 15 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:433]: `some` is a less surprising name for the same thing.

**Comment by cblp** [comment]
bors merge
